index.js

```diff
@@ -45,115 +82,192 @@ const propertiesISOEquivalent = {
+  }
 
-    return value + propertiesISOEquivalent[property]
+  if (currentValue < 1000000) {
+    return currentValue / 1000
   }
+
+  // slow path
+  const remainder = currentValue.toString().length - 3
+  return currentValue / Math.pow(10, remainder)
 }
 
-const NUMBER = '([+-]?\\d+)'
-const YEAR = `${NUMBER}\\s+years?`
-const MONTH = `${NUMBER}\\s+mons?`
-const DAY = `${NUMBER}\\s+days?`
-// NOTE: PostgreSQL automatically overflows seconds into minutes and minutes
-// into hours, so we can rely on minutes and seconds always being 2 digits
-// (plus decimal for seconds). The overflow stops at hours - hours do not
-// overflow into days, so could be arbitrarily long.
-const TIME = '([+-])?(\\d+):(\\d\\d):(\\d\\d(?:\\.\\d{1,6})?)'
-const INTERVAL = new RegExp(
-  '^\\s*' +
-    // All parts of an interval are optional
-    [YEAR, MONTH, DAY, TIME].map((str) => '(?:' + str + ')?').join('\\s*') +
-    '\\s*$'
-)
-
-// All intervals will have exactly these properties:
-const ZERO_INTERVAL = Object.freeze({
-  years: 0,
-  months: 0,
-  days: 0,
-  hours: 0,
-  minutes: 0,
-  seconds: 0,
-  milliseconds: 0.0
-})
-
-function parse (interval) {
+function parse (instance, interval) {
   if (!interval) {
-    return ZERO_INTERVAL
-  }
-
-  const matches = INTERVAL.exec(interval) || []
-
-  const [
-    ,
-    yearsString,
-    monthsString,
-    daysString,
-    plusMinusTime,
-    hoursString,
-    minutesString,
-    secondsString
-  ] = matches
-
-  const timeMultiplier = plusMinusTime === '-' ? -1 : 1
-
-  const years = yearsString ? parseInt(yearsString, 10) : 0
-  const months = monthsString ? parseInt(monthsString, 10) : 0
-  const days = daysString ? parseInt(daysString, 10) : 0
-  const hours = hoursString ? timeMultiplier * parseInt(hoursString, 10) : 0
-  const minutes = minutesString
-    ? timeMultiplier * parseInt(minutesString, 10)
-    : 0
-  const secondsFloat = parseFloat(secondsString) || 0
-  // secondsFloat is guaranteed to be >= 0, so floor is safe
-  const absSeconds = Math.floor(secondsFloat)
-  const seconds = timeMultiplier * absSeconds
-  // Without the rounding, we end up with decimals like 455.99999999999994 instead of 456
-  const milliseconds = Math.round(timeMultiplier * (secondsFloat - absSeconds) * 1000000) / 1000
-  return {
-    years,
-    months,
-    days,
-    hours,
-    minutes,
-    seconds,
-    milliseconds
+    return
+  }
+
+  position.value = 0
+
+  let currentValue
+  let nextNegative = 1
+
+  while (position.value < interval.length) {
+    const char = interval[position.value]
+
+    if (char === '-') {
+      nextNegative = -1
+      position.value++
+      continue
+    } else if (char === '+') {
+      position.value++
+      continue
+    } else if (char === ' ') {
+      position.value++
+      continue
+    } else if (char < '0' || char > '9') {
+      position.value++
+      continue
+    } else {
+      currentValue = readNextNum(interval)
+
+      if (interval[position.value] === ':') {
```

Comment:
Charcodes are slightly faster, you can compile the charcode of the char you want to compare against, and use it to validate the string